File: factorial/solutions/FactorialSolution.scala
package solutions

object FactorialSolution {

  def main(args: Array[String]) = {
    val input = io.Source.stdin.bufferedReader.readLine()
    println(factorial(input.toInt))
  }

  def factorial(n: Int): BigInt = {
    
    def calc(acc: BigInt, n: Int): BigInt =
      if (n == 0) acc
      else calc(acc * n, n - 1)

    calc(1, n)
  }

}

File: factorial/solutions/FactorialSolution.java
package solutions

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.math.BigInteger;


public class FactorialSolution {

  public static void main(String[] args) throws Exception {
    // read STDIN 
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BigInteger n = BigInteger.valueOf(Integer.parseInt(br.readLine()));
    // write STDOUT
    System.out.println(factorial(n, BigInteger.ONE));
  }

  /*
   *  Factorial of n 
   *  @param  n   input
   *  @param  acc accumulator (tail call optimization)
   *  @return n!   
   */
  public static BigInteger factorial(BigInteger n, BigInteger acc) {
    
    if (n.equals(BigInteger.ZERO))
      return acc;
    else
      return factorial(n.subtract(BigInteger.ONE), n.multiply(acc));      
  }    
}

